File: backend/app/schemas/transactions.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime

# ── OCR 端點 Schema ────────────────────────────────────────────

class OcrResultOut(BaseModel):
    """OCR 辨識結果回應 Schema"""
    amount: Optional[float] = None
    currency: str = "TWD"
    merchant_name: Optional[str] = None
    transaction_date: Optional[str] = None
    channel_name: Optional[str] = None
    category: Optional[str] = None
    confidence: str = "low"
    raw_text: Optional[str] = None

# ── 交易記帳 Schema ────────────────────────────────────────────

class TransactionCreate(BaseModel):
    """新增交易記帳請求 Schema"""
    user_card_id: int
    channel_name: str
    card_mode: Optional[str] = None
    category: str = "其他"
    merchant_name: Optional[str] = None
    original_amount: float
    currency: str = "TWD"
    source_type: str = "MANUAL"   # QR_CODE / AI_RECEIPT_VISION / AI_SCREENSHOT_VISION / MANUAL
    card_mode: Optional[str] = None
    transacted_at: Optional[str] = None  # ISO datetime 字串，若為空則使用當下時間

class TransactionOut(BaseModel):
    """交易記帳回應 Schema"""
    id: int
    user_card_id: int
    channel_name: str
    card_mode: Optional[str] = None
    category: str
    merchant_name: Optional[str]
    original_amount: float
    currency: str
    ntd_amount: float
    exchange_rate: float
    earned_cashback_ntd: float
    source_type: str
    card_mode: Optional[str]
    transacted_at: str

    model_config = {"from_attributes": True}

File: mobile/src/api/transactions.ts
import { apiClient } from './client';

export interface OcrResult {
  amount: number | null;
  currency: string;
  merchant_name: string | null;
  transaction_date: string | null;
  channel_name: string | null;
  category: string | null;
  confidence: string;
  raw_text: string | null;
}

export interface TransactionCreate {
  user_card_id: number;
  channel_name: string;
  category: string;
  merchant_name?: string;
  original_amount: number;
  currency: string;
  source_type: string;
  transacted_at?: string;
}

export interface TransactionOut {
  id: number;
  user_card_id: number;
  channel_name: string;
  category: string;
  merchant_name: string | null;
  original_amount: number;
  currency: string;
  ntd_amount: number;
  exchange_rate: number;
  earned_cashback_ntd: number;
  source_type: string;
  transacted_at: string;
}

/** 上傳收據/截圖圖片給 AI 辨識 */
export const uploadImageForOcr = async (
  imageUri: string,
  type: 'receipt' | 'screenshot',
): Promise<OcrResult> => {
  const formData = new FormData();
  const filename = imageUri.split('/').pop() ?? 'image.jpg';
  formData.append('file', {
    uri: imageUri,
    name: filename,
    type: 'image/jpeg',
  } as any);

  const endpoint = type === 'receipt' ? '/ocr/receipt' : '/ocr/screenshot';
  const res = await apiClient.post<OcrResult>(endpoint, formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
    timeout: 30000,
  });
  return res.data;
};

/** 新增交易記帳 */
export const createTransaction = async (payload: TransactionCreate): Promise<TransactionOut> => {
  const res = await apiClient.post<TransactionOut>('/transactions', payload);
  return res.data;
};

/** 取得交易明細列表 */
export const fetchTransactions = async (
  user_card_id?: number,
  start_date?: string,
  end_date?: string
): Promise<TransactionOut[]> => {
  const params: any = {};
  if (user_card_id) params.user_card_id = user_card_id;
  if (start_date) params.start_date = start_date;
  if (end_date) params.end_date = end_date;
  
  const res = await apiClient.get<TransactionOut[]>('/transactions', { params });
  return res.data;
};
